### PR TITLE
BufferGeometry addGroup and clearGroups methods

### PR DESCRIPTION
**Pull request by trusktr**

### CODE REVIEW

**Comment by ZoeDreams** [comment]
made BufferGeomteryGroup array strong typed as an Array<T> and not generic []

extended the clear test to make sure we can add items to the new group.

**Review by ZoeDreams — CHANGES_REQUESTED** [review]
okay so i made the array strong typed rather then implicit. Extended some clear tests..

also noticed that materialIndex was typed as f32 in faces.. this needs to be an integer.

File: src/as/core/BufferGeometry.ts [hunk 1]
```diff
@@ -71,7 +80,7 @@ export class BufferGeometry extends EventDispatcher {
 	morphAttributes: Map<string, BufferAttribute[]> = new Map()
 	//^ per BufferGeometryLoader.js in the original three.js, geometry.morphAttributes[key] is loaded with array of BufferAttributes
 
-	groups: BufferGeometryGroup[] = []
+	groups: BufferGeometryGroup[] = new Array<BufferGeometryGroup>()
```
Comment (trusktr) [inline src/as/core/BufferGeometry.ts:83]: `groups: BufferGeometryGroup[] = []` is the same thing, but less characters.

When TS (and AS) sees `[]` being assigned to an LHS that has the type declaration, then the RHS automatically has that type. It is type safe. It's just syntax short hand.

File: src/as/core/BufferGeometry.ts [hunk 2]
```diff
@@ -150,17 +161,24 @@ export class BufferGeometry extends EventDispatcher {
...
 
-	// clearGroups(): void {
-	// 	this.groups = []
-	// }
+	/**
+	 * clears buffer geometery group arrays by allocating a new empty array in place. This is required
+	 * to properly invoke the GC within the heap stack. 
+	 */
+	clearGroups(): void {
+		this.groups = new Array<BufferGeometryGroup>()
```
Comment (trusktr) [inline src/as/core/BufferGeometry.ts:180]: Using `this.groups.length = 0` prevents making a new object (one less object to GC later), and accomplishes the same goal (release all items in the array).

File: src/as/core/Face3.ts [hunk 3]
```diff
@@ -33,12 +36,12 @@ export class Face3 {
...
 	 */
 	constructor(
-		public a: f32,
-		public b: f32,
-		public c: f32,
+		public a: i32,
+		public b: i32,
+		public c: i32,
 		normals: Array<Vector3> = [],
 		colors: Array<Color> = [],
```
Comment (trusktr) [inline src/as/core/Face3.ts:43]: Here's an example, these array literals will have the type of the LHS.

File: src/as/core/BufferGeometry.ts (hunk 1 above)
Comment (trusktr) [reply src/as/core/BufferGeometry.ts:83]: In plain JS/TS both are the exact same (at runtime, compile time, and type check time). But dcode [answered here](https://discord.com/channels/721472913886281818/721497900932137090/787930839861821460) that using array literals is nice because it will try to optimize into a `StaticArray` under the hood (still appears as an `Array` to the code writer), and fall back to regular `Array` for dynamic length if something like `array.push` is used.

So basically if we use the literal format, we may sometimes get an optimization. This is good to know!

This means, for things like the `elements` array inside of `Matrix4` that contains 16 numbers, the length will never change, so initializing it as `[n1, n2, ..., n16]` will automatically make it a `StaticArray` without all the dynamic stuff that `Array` has. 

Neat!

Comment (trusktr) [reply src/as/core/BufferGeometry.ts:83]: In our case here, we are making empty `[]` arrays, so they will end up as `Array` because we need to expand their size later.

But TLDR, we can use `[]` for syntax short-hand, without penalty. If your IDE was showing any type error relating to `any[]`, we should see what is up with that. It works well on my end.

Comment (jtenner) [reply src/as/core/BufferGeometry.ts:83]: StaticArray is only used when you specify it because it has a different memory layout than Array. Using `[]` instead of `new Array()` is nicer though.

Comment (ZoeDreams) [reply src/as/core/BufferGeometry.ts:83]: good to know thanx

File: src/as/core/BufferGeometry.ts (hunk 2 above)
Comment (ZoeDreams) [reply src/as/core/BufferGeometry.ts]: neat trick.. updat3d.

**Comment by ZoeDreams** [comment]
completed the updates requested.

**Review by ZoeDreams — APPROVED** [review]
looks good. clean.
